itou/siae_evaluations/models.py

```diff
@@ -312,8 +313,42 @@ def __str__(self):
     @property
     def state(self):
         # property in progress, new conditionnal state will be added further
+        if self.evaluated_eligibility_diagnoses.exists():
+            return evaluation_enums.EvaluationJobApplicationsState.PROCESSING
         return evaluation_enums.EvaluationJobApplicationsState.PENDING
 
+    def save_selected_criteria(self, cleaned_keys=None, changed_keys=None):
+        # cleaned_keys are checked fields when form is submitted.
+        #    It contains new AND prexistant choices.
+        # changed_keys are fields which status has changed when form is submitted :
+        #    from unchecked to checked
+        #    from checked to unchecked
+        #    It contains new AND removed choices.
+
+        if changed_keys:
+            if cleaned_keys is None:
+                cleaned_keys = []
+
+            with transaction.atomic():
+
+                EvaluatedEligibilityDiagnosis.objects.filter(
+                    pk__in=(
+                        eval_diag.pk
+                        for eval_diag in self.evaluated_eligibility_diagnoses.all()
+                        if eval_diag.administrative_criteria.key in list(set(changed_keys) - set(cleaned_keys))
+                    )
+                ).delete()
+
+                EvaluatedEligibilityDiagnosis.objects.bulk_create(
+                    [
+                        EvaluatedEligibilityDiagnosis(
+                            evaluated_job_application=self, administrative_criteria=criterion
+                        )
+                        for criterion in AdministrativeCriteria.objects.all()
+                        if criterion.key in list(set(changed_keys).intersection(cleaned_keys))
+                    ]
+                )
```

Comment:
ok, mais deux choses métonnent un peu.

- peut etre que on devrait juste s'assurer que ce `save_slected_criteria` n'est pas appelé si `changed_keys` est vide et faire ça avant l'appel, ça evitera d'avoir juste toute la fonction dans un gros `if` ce qui est un _code smell_
- je suis etonné que on ne puisse pas partir de la `job_application` (donc le `self`) pour faire un truc du genre `self.foobar.diagnoses_set.delete()` plutot que de passer par un chemin détourné avec les `pk__in`
- le `list(set)` est inutile, tu peux itéret sur un set

Reply:
ok sur le `save_selected_criteria`.
ok sur le `list` inutile.
Sur le point 2, je vois moins. L'objectif est de ne supprimer que ceux qui ont été décochés dans le formulaire, donc présent dans `changed_keys`et absent de `cleaned_keys`

Reply:
OK compris pour le point 2. Le if est toujours là.

Reply:
good bye `if` :fire: